SVHN dataset for torchvision

uridah opened the pull request:
In reference to https://github.com/pytorch/vision/issues/59
cc: @mjpieters

soumith commented:
@uridah do you have a notebook that checks for sanity of this code, like the notebook that loads some CIFAR10 images and displays them:
https://github.com/pytorch/vision/blob/master/test/sanity_checks.ipynb

uridah commented:
@soumith I don't but I can create one.

mjpieters commented:
I don't know if PyTorch follows PEP8 as a styleguide? If so, you way want to run [flake8](http://flake8.pycqa.org/en/latest/) on your code and fix the issues that finds; I see a few whitespace issues in it. :-) If not, I'll leave it to the project owners to set the requirements.

soumith commented:
thanks @mjpieters . We have a LINT check as part of the contbuild, and it is failing:

https://travis-ci.org/pytorch/vision/builds/209884377

Locally doing:
```
pip install flake8
flake8 .
``
will show exact LINT errors you need to fix @uridah

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,161 @@
+                raise RuntimeError('Dataset not found or corrupted.' +
+                               ' You can use download=True to download it')
+            
+            self.train_data = []
```
those two lines are not necessary, as they are assigned just below

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,161 @@
+        self.dataset = dataset  # training set or test set or extra set
+        
+        # download and load the data 
+        if self.dataset=='train':
```
Maybe you could condensate these code-paths into a single one?
The only difference that I see is that you change the values of `url`, `filename` and `file_md5` for each set.
Maybe you could instead have a dictionary for each one of those fields, where the key is the set type (`train` / `extra`, `test`), and the values are the different urls, like done in [here](https://github.com/pytorch/vision/pull/63/files#diff-629efaeb461f617e4cdba4ef6b18222eR16)? Then you could have 
```python
self.url = url[dataset]
```

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,161 @@
+
+    def __len__(self):
+        if self.dataset == 'train':
+            return 73257
```
maybe it's better to just return `len(self.train_data)`?

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,161 @@
+        self.root = root
+        self.transform = transform
+        self.target_transform = target_transform
+        self.dataset = dataset  # training set or test set or extra set
```
maybe `dataset` is a misleading name, and something like `split` would be better?

fmassa commented:
@uridah I made some small inline comments (very minor and are only style changes). The PR looks good, thanks!

uridah commented:
@fmassa Your comments very very useful and really helped concise the code
@soumith I added a new notebook named sanity_checks1.ipynb which is basically same as the sanity_checks.ipynb notebook, but in addition, is also calling SVHN datasets for transformations https://github.com/uridah/vision/blob/master/test/sanity_checks1.ipynb
Also, ran the code through flake8
@mjpieters

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,117 @@
+            # reading(loading) mat file as array
+            loaded_mat = sio.loadmat(os.path.join(root, self.filename))
+
+            if self.split != 'test':
```
I think we don't need to separate between `self.train_data` and `self.test_data`, and just `self.data` is fine.
This way you avoid this condition and the one in `__getitem__` as well.

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,117 @@
+                self.test_labels = loaded_mat['y']
+                self.test_data = np.transpose(self.test_data, (3, 2, 1, 0))
+        else:
+            print ("Wrong dataset entered! Please use split=train or split=extra or split=test")
```
It might be better to raise an error in here.
Maybe something like
```python
raise ValueError("Wrong dataset entered! Please use split=train, split=extra or split=test")
```

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,117 @@
+        self.target_transform = target_transform
+        self.split = split  # training set or test set or extra set
+
+        if self.split in self.split_list:
```
I think it might be better to have a condition in the other way around. Something like
```python
if self.split not in self.split_list:
    raise ValueError("...")

# rest of the code goes here, without the extra identation
```

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,117 @@
+            print ("Wrong dataset entered! Please use split=train or split=extra or split=test")
+
+    def __getitem__(self, index):
+        if self.split == 'train' or self.split == 'extra':
```
If you remove the difference between `self.train_data` and `self.train_labels` and add only `self.data` and `self.labels`, this condition can be removed and the code is a bit simpler

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,117 @@
+        return img, target
+
+    def __len__(self):
+        return len(self.train_data)
```
Another reason why it might be better to merge `train_data` into `data` only. This currently doesn't work in the `test` case.

fmassa commented:
Hi @uridah 
I made some more comments. The code is looking very nice!

uridah commented:
Thanks @fmassa. Updated according to your suggestions

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,109 @@
+
+        if self.split not in self.split_list:
+            raise ValueError('Wrong split entered! Please use split=train or split=extra or split=test')
+        else:
```
One last thing, this `else` (and the idented code) is not needed anymore, because of the `raise` in the line just above.

fmassa commented on torchvision/datasets/svhn.py:
```diff
@@ -0,0 +1,109 @@
+            self.download()
+
+        if not self._check_integrity():
+                raise RuntimeError('Dataset not found or corrupted.' +
```
not important: the indentation in here is off

fmassa commented:
About the `check_sanity1.pynb` file, maybe you could integrate your changes into the `check_sanity.pynb` file? What do you think?

uridah commented:
I will, if you guys think it's good enough.

fmassa commented:
@uridah another thing, it would be great if you could add an entry in the doc in `README.md`, like the one in `Cifar10` for example. Maybe an example in the doc would be enough?
cc @soumith

uridah commented:
here is the pull request for the change in documentation: https://github.com/pytorch/vision/pull/104
@fmassa

fmassa commented:
Hi @uridah 
So, looking at the notebook that you added, I have the impression that the numbers are rotated. Maybe the `x` and `y` axis in the dataset are inverted, and you need to transpose them?
Also, could you please fix the indentation in the line I commented?
Once those are fixed, I think there is no need for the notebook (but @soumith knows better), so could you please remove it?
After that, the PR is good to be merged!

uridah commented:
@fmassa as it turns out I needed it to transpose along 3,2,0,1 axis instead of 3,2,1,0. I have fixed that and updated the indentation and sanity_checks1.ipynb. Please have a look and let me know if anything else needs to be changed.
Also, I created a separate request to update the documentation but now I have merged it with this one.

soumith commented:
Thanks Uridah, as you saw the last 4 commits, I made some minor changes to your PR. But it looked great. Merged into master now!!!

vabh commented:
Hello,

It seems that the labels returned are in the range `1-10`  (the data set assigns class `10` to the digit `0` and class `d` for all other digits, `d` [ http://ufldl.stanford.edu/housenumbers/ see section `overview` ]).

Given that some of the loss functions (CELoss, NLLLoss) expect the class labels to be in the range `[0, C-1]`, shouldn't that be reflected here as well?

Another thing is that the the returned `labels` are of type ByteTensor and of size `batchSize x 1`. Again, CELoss, etc, expect it to be a 1d `LongTensor`.

I can make a PR if the current behaviour is inconsistent and should be made similar to the other data sets, for example as in CIFAR10

soumith commented:
@vabh if you could make a PR to reflect this, that'd be great. Thanks.

vabh commented:
@soumith I made one here: #194
